l2-contracts/contracts/ConsensusAuthority.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import "./ValidatorRegistry.sol";
+import "./AttesterRegistry.sol";
+
+contract ConsensusAuthority {
+    // Owner of the contract (i.e., the governance contract).
+    address public owner;
+    // The validator registry instance.
+    ValidatorRegistry public validatorRegistry;
+    // The attester registry instance.
+    AttesterRegistry public attesterRegistry;
+
+    // A map of owningPubKey (a cold key L2 address) => indexes on registries.
+    mapping(address => RegistryIndexes) public nodes;
+
+    struct RegistryIndexes {
+        // Index in the validator registry.
+        uint256 validatorIdx;
+        // Index in the attester registry.
+        uint256 attesterIdx;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Unauthorized: onlyOwner");
+        _;
+    }
+
+    modifier onlyOwnerOrNodeOwner(address owningPubKey) {
+        require(msg.sender == owner || msg.sender == owningPubKey, "Unauthorized: onlyOwnerOrNodeOwner");
+        _;
+    }
+
+    constructor(address _owner) {
+        owner = _owner;
+        validatorRegistry = new ValidatorRegistry(address(this));
+        attesterRegistry = new AttesterRegistry(address(this));
+    }
+
+    // Adds a new node to the registries.
+    function add(
+        address owningPubKey,
+        uint256 weight,
+        bytes calldata validatorPubKey,
+        bytes calldata validatorPoP,
+        bytes calldata attesterPubKey
```

Comment:
I thought validators and attesters can be independent. Even if they are attached to a node, and they have a key don't use, shouldn't they have two weights to say "this node has 0 weight as a validator but W as an attester"?

Reply:
Should have different weights, correct.

Reply:
Fixed.